mesonbuild/_pathlib.py

```diff
@@ -0,0 +1,43 @@
+# Copyright 2021 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import pathlib
+import os
+
```

Comment:
It does not seem possible to do this without monkey patching, so that's what we need to do then. However the implementation needs some fixes:

- the name should not be `_pathlib` as leading underscores in names are confusing and can lead to problems (as can be seen with the CI), instead it should be something more dangerous sounding like `pybugXXXXworkaround`
- The file should have a comment at the top explaining what it is, why it's needed, have links to the relevant bug reports, a notice that no new code or functionality should be added to it and also when the workaround can be removed.

Reply:
Done.

Reply:
Well, I guess renaming the file did not fix the `1 for Module imports itself` warning...

Reply:
That seems like an LGTM bug then. Let's just rename it `pathlibworkaround` and call it a day.

Reply:
So, do you want me to rename it again, or does the current name work?

Reply:
According to the web ui the module name is `zzz_pathlibworkaround` which is not a great name ...

Reply:
But, then it looks like a regular meson module and blends in with the rest of them, which goes against the point of making it look scary. Also, we already have the `_typing` module...

Would you be ok with renaming it to `zzzpathlibworkaround`? Also which web ui is complaining about the name?

Reply:
I believe LGTM is complaining that whatever the name of your module, it's being inserted into `sys.modules['pathlib']` while at the same time importing `pathlib`.

Which I still think is a bad idea, when we have _typing we import from _typing rather than inserting it into `sys.modules['typing']`.

Reply:
> Also which web ui is complaining about the name?

No system complains, this is 100% my own esthetical eye thing. A name like `zzz_anything` reads like "a placeholder name thrown in to test something which will be replaced with the real name later". I don't really have strong opinions about the naming, it just seemed strange to me.

Reply:
The difference to `_typing` is that it is *never* actually imported in production. It is only used in `if T.TYPE_CHECKING:` blocks.

If we do the same with `pathlib` and `_pathlib` we have to replace all `import pathlib` and `from pathlib import *`, remember to always to use `_pathlib` (which can result in non-obvious errors otherwise). And once we remove the workaround we have to revert all these changes again... IMHO, at some point one dirty hack is more convenient and developer-friendly than renaming 50 different strings and enforcing that.

@jpakkane Would you then be OK, to revert to renaming it to `_pathlib` since LGTM is just broken in this regard? I really don't want it to be sorted in the middle of the other meson modules and have it blend in and look more legitimate than it is.

Reply:
Fine by me.

Reply:
Actually, would `import _pathlib as pathlib` work better?

Reply:
If we do `import _pathlib as pathlib` or `from _pathlib import Path`, then we have to do it *everywhere* and *consitently*, which is why I wanted to do the `sys.modules['typing']` hack in the first place. This definetely is not the most elegant solution, but the soultion with the smallest git diff (not that this would make a difference after we went for fixing `encoding` on the call site...) and the least maintenance burden.

Reply:
Let's go with the current one then, it can be always fixed later. One thing to note, though, is that since this issue only occurs on Windows, the monkeypatching should probably also be done on Windows.

Reply:
*If* we do `import _pathlib` we need *another* custom linter for a semi temporary fix/hack to ensure that we don't mix `_pathlib` and `pathlib`

Reply:
Done.